Blog Landing Page - featured video post

danielfmiranda opened the pull request:
Closes #8685 
Related PRs/issues #


### Screenshots:

CMS
---
This is what the "featured video post" section appears like in the CMS. This field is optional, meaning if there is no featured post set, the front end wont render at all. However, if the user chooses to set one, both the page and the video URL are required.
<img width="1130" alt="Screen Shot 2022-06-15 at 6 12 48 PM" src="https://user-images.githubusercontent.com/18314510/173970884-55a28470-be45-4c20-8984-e7b0b9c8b0f3.png">


Front End (With Vimeo Link)
---
Desktop:
<img width="1130" alt="Screen Shot 2022-06-15 at 6 12 48 PM" src="https://user-images.githubusercontent.com/18314510/173971002-b4bc2f4c-7d30-4e9f-aac2-af9014b40a79.png">
Mobile:
<img width="350" alt="Screen Shot 2022-06-15 at 6 12 48 PM" src="https://user-images.githubusercontent.com/18314510/173971009-c20a4ee9-e402-401e-b258-1e9af6fe963b.png">


Front End (With Youtube Link)
---
Desktop:
<img width="1130" alt="Screen Shot 2022-06-15 at 6 12 48 PM" src="https://user-images.githubusercontent.com/18314510/173971186-d02519cb-5a0c-43fb-9c46-01a7f3b26c78.png">
Mobile:
<img width="350" alt="Screen Shot 2022-06-15 at 6 12 48 PM" src="https://user-images.githubusercontent.com/18314510/173971191-d362fdab-67f9-4d88-94b0-a425ad927750.png">

danielfmiranda commented on network-api/networkapi/wagtailpages/templates/wagtailpages/blog_index_page.html:
```diff
@@ -22,6 +22,15 @@
 
+{% block featured_video_post %}
+  {% if not filtered %}
+    {%  for video_post in page.specific.featured_video_post.all %}
```
@tbrlpld Drawing inspiration from a previous ticket that we worked together on, instead of using an if statement I am just looping through the list, even though it is limited to 1 choice. So if there is not one set in the CMS, this field wont render at all

danielfmiranda commented on network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/blog_index_featured_video_post.html:
```diff
@@ -0,0 +1,43 @@
+  <div class="tw-grid tw-grid-cols-12 tw-mb-7">
+    <div class="tw-pr-[36px]  tw-py-3  tw-flex tw-flex-col tw-col-span-12 large:tw-col-span-4 large:tw-py-0">
+      <div class="tw-flex tw-flex-wrap">
+        {% with topics=blog_page.topics.all topic_classes="tw-h6-heading tw-text-blue-80 tw-mr-2 tw-my-1" %}
```
Follow up ticket to componentize this: https://github.com/mozilla/foundation.mozilla.org/issues/8980

tbrlpld replied:
Yea good call 👍

danielfmiranda commented on package-lock.json:
```diff
@@ -1510,6 +1510,12 @@
         "unist-util-find-all-after": "^3.0.2"
       }
     },
+    "@tailwindcss/aspect-ratio": {
```
Hi @fessehaye! It turns out Tailwind has a aspect-ration package that we can use instead of bootstraps. Just wanted to double check if I was adding it to package-lock.json correctly 😄 thanks!

fessehaye reviewed (changes requested):
If we are just using 16/9 ratios, we should just use the built in aspect ratio classes and remove the plugin from the package.json 

https://tailwindcss.com/docs/aspect-ratio#setting-the-aspect-ratio

`tw-aspect-video` could have replaced all of the `aspect-{}` classes you used

tbrlpld commented on network-api/networkapi/wagtailpages/pagemodels/blog/blog_index.py:
```diff
@@ -44,6 +44,38 @@ def __str__(self):
+        help_text='For YouTube: go to your YouTube video and click “Share,” '
+                  'then “Embed,” and then copy and paste the provided URL only. '
+                  'For example: https://www.youtube.com/embed/3FIVXBawyQw '
+                  'For Vimeo: Log into Vimeo using 1Password '
```
**Suggestion** I wonder if we could force a line break in the help text between the paragraph on YouTube and the one on Vimeo.

tbrlpld commented on network-api/networkapi/wagtailpages/pagemodels/blog/blog_index.py:
```diff
@@ -44,6 +44,38 @@ def __str__(self):
+        related_name='+'
+    )
+    video_url = models.URLField(
+        help_text='For YouTube: go to your YouTube video and click “Share,” '
```
```suggestion
        help_text='For YouTube: Go to your YouTube video and click “Share,” '
```

tbrlpld commented on network-api/networkapi/wagtailpages/templates/wagtailpages/blog_index_page.html:
```diff
@@ -1,6 +1,6 @@
 {% extends "./index_page.html" %}
 
-{% load wagtailcore_tags wagtailimages_tags i18n %}
+{% load wagtailcore_tags wagtailimages_tags i18n localization homepage_tags blog_tags static %}
```
**Question** Is this necessary anymore? Feels like it was leftover from refactoring.

danielfmiranda replied:
Updated!

tbrlpld commented on network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/blog_index_featured_video_post.html:
```diff
@@ -0,0 +1,43 @@
+          {% endfor %}
+        {% endwith %}
+        <img src="{% static "_images/video-icon.svg" %}" alt="{% trans "video" context "icon description for screen readers" %}" class="tw-h-4 tw-w-4 tw-mr-2 tw-my-1">
+        <span class="tw-h6-heading tw-text-gray-40 tw-my-1"> {{ blog_page.first_published_at|date:"DATE_FORMAT" }} </span>
```
**Question** Can we achieve this spacing with padding rather than actual whitespace?

danielfmiranda replied:
Hi @tbrlpld! I am not sure if I 100% understand your question, however, I updated the "my-1" classes to "py-1" classes to use padding instead of margin for these items. The only one I was not able to update was the image, as if I updated it to use padding, the SVG would shrink.

tbrlpld replied:
Sorry if I was not clear enough here. I mean the whitespace in the span tag. Specifically between the span tag and the curly braces. 

My suggestion would have been to remove the white space and instead rely on padding (or margin) on the span element to recreate the spacing (if necessary).
```
        <span class="tw-h6-heading tw-text-gray-40 tw-my-1">{{ blog_page.first_published_at|date:"DATE_FORMAT" }}</span>
```

tbrlpld commented on network-api/networkapi/wagtailpages/templates/wagtailpages/index_page.html:
```diff
@@ -59,7 +62,7 @@ <h3 class="tw-body result-info">
     {% endif %}  {# End of IF for filtered page content _after_ starting a new row. #}
       {% block subcontent %}
         {% include "./fragments/entry_cards.html" %}
-      {% endblock %}
+    {% endblock %}
```
**Suggestion** The above two lines should be out-dented too.

danielfmiranda replied:
Good catch! Updated

tbrlpld commented on network-api/networkapi/wagtailpages/templates/wagtailpages/blog_index_page.html:
```diff
@@ -22,6 +22,15 @@
   {% endif %}
 {% endblock %}
 
+{% block featured_video_post %}
+  {% if not filtered %}
+    {%  for video_post in page.specific.featured_video_post.all %}
+      {% include "./fragments/blog_index_featured_video_post.html" with blog_page=video_post.blog_page video_url=video_post.video_url %}
+    {% endfor %}
+  {% endif %}
+
+{% endblock %}
```
**Suggestion** The "featured video post" still fits the "featured" concept. Why not add it to the existing block above? You would not need to edit the base template in that case.

danielfmiranda replied:
Great suggestion Tibor, I updated it to be in the featured section as that makes more sense to me as well!

tbrlpld commented on network-api/networkapi/wagtailpages/pagemodels/blog/blog_index.py:
```diff
@@ -44,6 +44,38 @@ def __str__(self):
+        on_delete=models.CASCADE,
+        related_name='+'
+    )
+    video_url = models.URLField(
```
**Suggestion** I believe the copy process could be simplified for editors by using the [`embed` template tag](https://docs.wagtail.org/en/stable/advanced_topics/embeds.html#embed-tag) to convert the oembed URL to the Video URL.

danielfmiranda replied:
Updated to use oembed. great suggestion!

tbrlpld reviewed (changes requested):
This is coming a long nicely @danielfmiranda Just a few small suggestions.

danielfmiranda commented:
Thanks for the great feedback @tbrlpld @fessehaye!!

Simon, I have implemented:
- the tailwind-aspect-[] class instead of having to use the package that I introduced earlier, I also removed all instances of that package and confirm that everything is working well

Tibor, I have:
- Updated help text like requested, unfortunately I could not find a timely way to put a line break in between both sentences though.
- Removed unneccesary tag imports 
- Moved the featured video section from its own block to the {% featured %} block
- Updated the template to use the {% embed %} block for youtube videos. I left vimeo alone however, as that is being rendered in a custom player vs a vimeo embed.

I have re-requested review from you both for whenever you have a free moment. Thanks so much!

fessehaye reviewed (changes requested):
Use `tw-aspect-video` instead of `tw-aspect-[16/9]`

https://tailwindcss.com/docs/aspect-ratio

tbrlpld commented on network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/blog_card.html:
```diff
@@ -16,7 +16,7 @@
 
 {% block published_date %}
-  <span class="tw-h6-heading tw-text-gray-40 tw-my-1"> {{ page.first_published_at|date:"DATE_FORMAT" }} </span>
+  <span class="tw-h6-heading tw-text-gray-40 tw-py-1 tw-mb-0"> {{ page.first_published_at|date:"DATE_FORMAT" }} </span>
```
Sorry for not making this clear before. I meant this:

```suggestion
  <span class="tw-h6-heading tw-text-gray-40 tw-py-1 tw-mb-0">{{ page.first_published_at|date:"DATE_FORMAT" }}</span>
```

If this works.

tbrlpld commented on network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/blog_card_topics.html:
```diff
@@ -4,8 +4,8 @@
+    <a class="tw-h6-heading tw-text-blue-80 tw-mr-2 tw-py-1 tw-mb-0" href="{% localizedroutablepageurl parent_page "entries_by_topic" topic.slug %}">{{ localized_topic }}</a>
     {% if not forloop.last %}
-      <span class="tw-h6-heading tw-text-blue-80 tw-mr-2 tw-my-1"> / </span>
+      <span class="tw-h6-heading tw-text-blue-80 tw-mr-2 tw-py-1 tw-mb-0"> / </span>
```
**Suggestion** Similarly here. I would suggest using margin or padding instead of whitespace in the HTML to create the spacing you want. 

```suggestion
      <span class="tw-h6-heading tw-text-blue-80 tw-mr-2 tw-py-1 tw-mb-0">/</span>
```

tbrlpld commented on network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/blog_index_featured_video_post.html:
```diff
@@ -1,42 +1,34 @@
-            {% endif %}
-          {% endfor %}
-        {% endwith %}
+        {% include "./blog_card_topics.html" with topics=blog_page.specific.topics.all%}
```
**Kudos** Nice. This is a lot more readable 👍

tbrlpld commented on network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/blog_index_featured_video_post.html:
```diff
@@ -1,42 +1,34 @@
+        {% include "./blog_card_topics.html" with topics=blog_page.specific.topics.all%}
         <img src="{% static "_images/video-icon.svg" %}" alt="{% trans "video" context "icon description for screen readers" %}" class="tw-h-4 tw-w-4 tw-mr-2 tw-my-1">
-        <span class="tw-h6-heading tw-text-gray-40 tw-my-1"> {{ blog_page.first_published_at|date:"DATE_FORMAT" }} </span>
+        <span class="tw-h6-heading tw-text-gray-40 tw-py-1"> {{ blog_page.first_published_at|date:"DATE_FORMAT" }} </span>
```
See above

```suggestion
        <span class="tw-h6-heading tw-text-gray-40 tw-py-1">{{ blog_page.first_published_at|date:"DATE_FORMAT" }}</span>
```

tbrlpld commented on network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/blog_index_featured_video_post.html:
```diff
@@ -1,42 +1,34 @@
           {% if 'player.vimeo.com' in video_url %}
-            <div id="featured-video-container" class="tw-group tw-aspect-w-16 tw-aspect-h-9 tw-h-full hover:tw-cursor-pointer">
-              <video class="featured-video tw-absolute tw-top-0 tw-left-0 tw-object-cover">
+            <div id="featured-video-container" class=" tw-relative tw-aspect-[16/9] tw-w-full tw-h-full hover:tw-cursor-pointer">
```
**Nitpick**

```suggestion
            <div id="featured-video-container" class="tw-relative tw-aspect-[16/9] tw-w-full tw-h-full hover:tw-cursor-pointer">
```

tbrlpld reviewed (approved):
Thanks for this @danielfmiranda This is coming nicely. Left comments trying to clarify my previous comment on the whitespace. I'm not sure that would work. Therefore regard it as optional.

fessehaye reviewed (approved):
thx for the changes!
